File: olive/engine/packaging/sample_code/ONNXModel/cs/code_sample.cs
/*
-------------------------------------------------------------------------
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
--------------------------------------------------------------------------
*/
using Microsoft.ML.OnnxRuntime;
using Newtonsoft.Json.Linq;

class Program
{
    static void UpdateSessOptions(SessionOptions sessOptions, JObject sessionOptions)
    {
        int? interOpNumThreads = null;
        if (sessionOptions.ContainsKey("inter_op_num_threads"))
        {
            interOpNumThreads = sessionOptions.Value<int?>("inter_op_num_threads");
        }
        int? intraOpNumThreads = null;
        if (sessionOptions.ContainsKey("intra_op_num_threads"))
        {
            intraOpNumThreads = sessionOptions.Value<int?>("intra_op_num_threads");
        }
        int executionMode = sessionOptions.Value<int>("execution_mode");
        int graphOptimizationLevel = sessionOptions.Value<int>("graph_optimization_level");
        JObject extraSessionConfig = sessionOptions.Value<JObject>("extra_session_config");

        if (interOpNumThreads.HasValue)
        {
            sessOptions.InterOpNumThreads = interOpNumThreads.Value;
        }
        if (intraOpNumThreads.HasValue)
        {
            sessOptions.IntraOpNumThreads = intraOpNumThreads.Value;
        }
        if (executionMode != -1)
        {
            if (executionMode == 0)
            {
                sessOptions.ExecutionMode = ExecutionMode.ORT_SEQUENTIAL;
            }
            else if (executionMode == 1)
            {
                sessOptions.ExecutionMode = ExecutionMode.ORT_PARALLEL;
            }
        }
        if (graphOptimizationLevel != -1)
        {
            sessOptions.GraphOptimizationLevel = (GraphOptimizationLevel)graphOptimizationLevel;
        }
        if (extraSessionConfig != null)
        {
            foreach (var kvp in extraSessionConfig)
            {
                sessOptions.AddSessionConfigEntry(kvp.Key, kvp.Value.ToString());
            }
        }
    }

    static void UpdateExecutionProvider(SessionOptions sessOptions, JArray executionProviderJson)
    {
        // Please note that Onnxruntime for C# has limited support for execution providers.
        // Refer to the Onnxruntime documentation for more information.
        // Please specify your device id to each provider if necessary
        string executionProvider = (string)executionProviderJson[0][0];
        JObject epOptionsJson = (JObject)executionProviderJson[0][1];
        Dictionary<string, string> epOptions = epOptionsJson.ToObject<Dictionary<string, string>>();

        if (executionProvider == "CUDAExecutionProvider")
        {
            OrtCUDAProviderOptions providerOptions = new OrtCUDAProviderOptions();
            providerOptions.UpdateOptions(epOptions);
            sessOptions.AppendExecutionProvider_CUDA(providerOptions);
        }

        if (executionProvider == "TensorrtExecutionProvider")
        {
            OrtTensorRTProviderOptions providerOptions = new OrtTensorRTProviderOptions();
            providerOptions.UpdateOptions(epOptions);
            sessOptions.AppendExecutionProvider_Tensorrt(providerOptions);
        }

        if (executionProvider == "OpenVINOExecutionProvider")
        {
            sessOptions.AppendExecutionProvider_OpenVINO();
        }

        if (executionProvider == "SNPEExecutionProvider")
        {
            sessOptions.AppendExecutionProvider("SNPE", epOptions);
        }

        if (executionProvider == "DmlExecutionProvider")
        {
            sessOptions.AppendExecutionProvider_DML();
        }
    }
    static void Main()
    {
        string path = @"c:\path\to\inference_config.json";
        string jsonString = File.ReadAllText(path);
        JObject j = JObject.Parse(jsonString);

        InferenceSession session;
        string modelPath = @"c:\path\to\model.onnx";
        if (j == null)
        {
            // Create inference session
            session = new InferenceSession(modelPath);
        }
        else
        {
            // Get execution_provider section
            JArray executionProvider = (JArray)j["execution_provider"];

            // Get session_options section
            JObject sessOpts = (JObject)j["session_options"];

            // Create inference configuration
            using SessionOptions options = new SessionOptions();
            UpdateSessOptions(options, sessOpts);
            UpdateExecutionProvider(options, executionProvider);

            // Create inference session
            session = new InferenceSession(modelPath, options);
        }

        // Run inference
        // Replace inputs, outputNames with actual variable names
        using var outputTensors = session.Run(<inputs>, <outputsNames>, new RunOptions());

        // Update output logic here
        var outputTensor = outputTensors.First().AsTensor<float>();

        session.Dispose();
    }
}

File: olive/engine/packaging/sample_code/ONNXModel/python/code_sample.py
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------
import json

import onnxruntime

# ruff: noqa


def run():
    # Load inference configuration json file
    with open("inference_config.json") as f:
        inference_settings = json.load(f)
    if inference_settings:
        session_options = inference_settings.get("session_options")
        execution_provider = inference_settings.get("execution_provider")
        use_ort_extensions = inference_settings.get("use_ort_extensions")

        # Onnxruntime inference session options
        sess_options = onnxruntime.SessionOptions()
        if use_ort_extensions:
            from onnxruntime_extensions import get_library_path

            sess_options.register_custom_ops_library(get_library_path())

        _update_sess_options(sess_options, session_options)
        session = onnxruntime.InferenceSession("model.onnx", sess_options, providers=execution_provider)
    else:
        # Use default inference session
        session = onnxruntime.InferenceSession("model.onnx")
    # Inference by session
    inputs = "input"  # input data
    output_names = "output"  # output names

    outputs = session.run([output_names], inputs)

    return outputs


def _update_sess_options(sess_options, session_options):
    inter_op_num_threads = session_options.get("inter_op_num_threads")
    intra_op_num_threads = session_options.get("intra_op_num_threads")
    execution_mode = session_options.get("execution_mode")
    graph_optimization_level = session_options.get("graph_optimization_level")
    extra_session_config = session_options.get("extra_session_config")
    if inter_op_num_threads:
        sess_options.inter_op_num_threads = inter_op_num_threads
    if intra_op_num_threads:
        sess_options.intra_op_num_threads = intra_op_num_threads
    if execution_mode:
        if execution_mode == 0:
            sess_options.execution_mode = onnxruntime.ExecutionMode.ORT_SEQUENTIAL
        elif execution_mode == 1:
            sess_options.execution_mode = onnxruntime.ExecutionMode.ORT_PARALLEL
    if graph_optimization_level:
        sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel(graph_optimization_level)
    if extra_session_config:
        for key, value in extra_session_config.items():
            sess_options.add_session_config_entry(key, value)
